### PR TITLE
Fix: Unit test fix and Bug fix

### PR DESCRIPTION
**Pull request by PinrLi**
Based on Issue #465
The current unit test and the semantic workflow code have a few errors. This PR fixes:
1. _discover_packages uses the operators_path incorrectly, fixed for general cases
2. get_package_commits didn't check if a commit is the initial commit, which might cause an edge case error
3. Several tests using incorrect paths for operators, ie: 
image
4. update_package_versions has a different logic between the code and the unit test for existing tags, modified the original code to match the unit test.
5. test_multiple_breaking_changes test needs more inspection on changing the test or the code. One of the messages was "feat: new feature without breaking change", while this message would be recognized as **Breaking** by _parse_conventional_commit. Not sure what is the best solution to this, skip for now.

Current Status
image

### CODE REVIEW

**Comment by PinrLi** [comment]
> hi @PinrLi thank you for your work, do give me sometime to review it
> 
> The test was written not properly earlier, so we should make changes to the test only and **NOT the script**, the script is working fine. We know that because we have written a lot of automated tests to make sure the script works.
> 
> Is it possible for you to undo the changes made to `scripts/semantic_release_workflow.py` in this PR? and modify the test accordingly, for instance, if you feel `update_package_versions()` has a different logic, then change the test function and not the script, or remove this function from the test.
> 
> Also great catching the issue in `test_multiple_breaking_changes`, we can fix the bug in `_parse_conventional_commit` later on

Sure I will work on it

File: scripts/semantic_release_workflow.py [hunk 1]
```diff
@@ -68,10 +68,11 @@ def _discover_packages(self):
         package_roots = ["feluda"]
 
         # Discover packages inside 'operators' directory using glob
-        operators_path = "operators"
+        operators_path = f"{self.repo_root}/operators"
```
Comment (aatmanvaidya) [inline scripts/semantic_release_workflow.py:71]: also if changes like this are absolutely important, then do let me know
we will have to modify the script in that case and not so much the test

because I understand how `f"{self.repo_root}/operators"` would be needed for the discover package function to find the folder in a `tmp` directory.

Comment (PinrLi) [reply scripts/semantic_release_workflow.py:71]: Yes I was just thinking about this, since the current implementation is hardcoded and would never check the tmp file, but all other changes could be reflected in the unit test file.

Comment (aatmanvaidya) [reply scripts/semantic_release_workflow.py:71]: right right that make's sense
so now you tell me what are some must have changes that need to be made

Comment (PinrLi) [reply scripts/semantic_release_workflow.py:71]: I think just this, and everything else I should be able to fix the test

**Comment by PinrLi** [comment]
Updated, fixed more path errors in test and in the workflow.
Also, fixed the test that works different from the code.

**Comment by aatmanvaidya** [comment]
perfect thanks, testing it out on my end too!
so the only change now in the semantic script is appending `{self.repo_root}` to the path's right?

**Comment by PinrLi** [comment]
> perfect thanks, testing it out on my end too! so the only change now in the semantic script is appending `{self.repo_root}` to the path's right?

That's right!

**Comment by aatmanvaidya** [comment]
hi @PinrLi I pulled the PR on my end, and ran the test like this
```sh
uv run -m unittest tests/scripts/test_semantic_release_workflow.py
```

I am getting the following errors
```sh
======================================================================
ERROR: test_get_tag_format (tests.scripts.test_semantic_release_workflow.TestPackageVersionManager)
Test the tag format generation.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/aatman/Aatman/Tattle/feluda/tests/scripts/test_semantic_release_workflow.py", line 342, in test_get_tag_format
    expected_tag = manager.packages["feluda"].get("pyproject_data", {}).get(
KeyError: 'feluda'

======================================================================
ERROR: test_tag_exists (tests.scripts.test_semantic_release_workflow.TestPackageVersionManager)
Test detection of existing tags.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/aatman/Aatman/Tattle/feluda/tests/scripts/test_semantic_release_workflow.py", line 362, in test_tag_exists
    self.assertTrue(manager.tag_exists(manager.packages["feluda"], "0.2.0"))
KeyError: 'feluda'

======================================================================
FAIL: test_determine_package_bump (tests.scripts.test_semantic_release_workflow.TestPackageVersionManager)
Test determination of version bump type based on commits.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/aatman/Aatman/Tattle/feluda/tests/scripts/test_semantic_release_workflow.py", line 289, in test_determine_package_bump
    self.assertEqual(
AssertionError: None != 'minor'

----------------------------------------------------------------------
Ran 21 tests in 0.712s

FAILED (failures=1, errors=2, skipped=2)
```

did I do something wrong?

**Comment by PinrLi** [comment]
> hi @PinrLi I pulled the PR on my end, and ran the test like this
> 
> ```shell
> uv run -m unittest tests/scripts/test_semantic_release_workflow.py
> ```
> 
> I am getting the following errors
> 
> ```shell
> ======================================================================
> ERROR: test_get_tag_format (tests.scripts.test_semantic_release_workflow.TestPackageVersionManager)
> Test the tag format generation.
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/aatman/Aatman/Tattle/feluda/tests/scripts/test_semantic_release_workflow.py", line 342, in test_get_tag_format
>     expected_tag = manager.packages["feluda"].get("pyproject_data", {}).get(
> KeyError: 'feluda'
> 
> ======================================================================
> ERROR: test_tag_exists (tests.scripts.test_semantic_release_workflow.TestPackageVersionManager)
> Test detection of existing tags.
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/aatman/Aatman/Tattle/feluda/tests/scripts/test_semantic_release_workflow.py", line 362, in test_tag_exists
>     self.assertTrue(manager.tag_exists(manager.packages["feluda"], "0.2.0"))
> KeyError: 'feluda'
> 
> ======================================================================
> FAIL: test_determine_package_bump (tests.scripts.test_semantic_release_workflow.TestPackageVersionManager)
> Test determination of version bump type based on commits.
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/aatman/Aatman/Tattle/feluda/tests/scripts/test_semantic_release_workflow.py", line 289, in test_determine_package_bump
>     self.assertEqual(
> AssertionError: None != 'minor'
> 
> ----------------------------------------------------------------------
> Ran 21 tests in 0.712s
> 
> FAILED (failures=1, errors=2, skipped=2)
> ```
> 
> did I do something wrong?

Seemed like I didn't fix these 3, can you retry my latest commit?
I got 
----------------------------------------------------------------------
Ran 21 tests in 1.216s

OK (skipped=2)

**Comment by aatmanvaidya** [comment]
@PinrLi it works now, will merge the PR soon!

**Comment by aatmanvaidya** [comment]
hi @PinrLi I have a [dummy repo](https://github.com/aatmanvaidya/feluda-versioning-check/) where I have written some [automated tests](https://github.com/aatmanvaidya/feluda-versioning-check/tree/main/automate_testing) to test out the semantic versioning workflow. 

Now I made a PR with a change in the `feluda/` folder ([commit](https://github.com/aatmanvaidya/feluda-versioning-check/commit/b68abf57dfcd8e5fe4001e0961cf26aa7ac74687)), and the script gave this error in the in when the [github action ran](https://github.com/aatmanvaidya/feluda-versioning-check/actions/runs/14399441415/job/40381726182)
```
Error discovering package at /home/runner/work/feluda-versioning-check/feluda-versioning-check/feluda: pyproject.toml not found in /home/runner/work/feluda-versioning-check/feluda-versioning-check/feluda
```

I think appending `self.repo_root` is causing the double repo name `feluda-versioning-check/feluda-versioning-check/feluda` and this is breaking the action. 

Apologies if I am being too nit-picky, since this workflow is very crucial part of `feluda`, I want to make sure all checks and automated tests pass

**Comment by PinrLi** [comment]
> hi @PinrLi I have a [dummy repo](https://github.com/aatmanvaidya/feluda-versioning-check/) where I have written some [automated tests](https://github.com/aatmanvaidya/feluda-versioning-check/tree/main/automate_testing) to test out the semantic versioning workflow.
> 
> Now I made a PR with a change in the `feluda/` folder ([commit](https://github.com/aatmanvaidya/feluda-versioning-check/commit/b68abf57dfcd8e5fe4001e0961cf26aa7ac74687)), and the script gave this error in the in when the [github action ran](https://github.com/aatmanvaidya/feluda-versioning-check/actions/runs/14399441415/job/40381726182)
> 
> ```
> Error discovering package at /home/runner/work/feluda-versioning-check/feluda-versioning-check/feluda: pyproject.toml not found in /home/runner/work/feluda-versioning-check/feluda-versioning-check/feluda
> ```
> 
> I think appending `self.repo_root` is causing the double repo name `feluda-versioning-check/feluda-versioning-check/feluda` and this is breaking the action.
> 
> Apologies if I am being too nit-picky, since this workflow is very crucial part of `feluda`, I want to make sure all checks and automated tests pass

I agree, I will try to resolve this

**Comment by PinrLi** [comment]
Can you try the current version if this works?

**Comment by PinrLi** [comment]
> @aatmanvaidya Can you try the current version if this works?

**Comment by aatmanvaidya** [comment]
hi @PinrLi I was testing on my end and faced an issue when we make any change to the root `pyproejct.toml` file, whenever any change is made to the root [`pyproejct.toml`](https://github.com/tattle-made/feluda/blob/main/pyproject.toml) file, the semantic version should be applied to `felulda`. 

So in my test, I made [this commit](https://github.com/aatmanvaidya/feluda-versioning-check/commit/5f4d696e021c7b5236c9ac507c83819b7e84ede3) to the pyproject.toml file - then the [github action gave this output](https://github.com/aatmanvaidya/feluda-versioning-check/actions/runs/14436724272/job/40478989436). 
Here is the output of the script
```sh
No changes found for /home/runner/work/feluda-versioning-check/feluda-versioning-check/feluda. Skipping version bump.
No changes found for /home/runner/work/feluda-versioning-check/feluda-versioning-check/operators/image_vec_rep_resnet. Skipping version bump.
No changes found for /home/runner/work/feluda-versioning-check/feluda-versioning-check/operators/vid_vec_rep_clip. Skipping version bump.
```

I think this issue is coming form these lines in the code - https://github.com/tattle-made/feluda/blob/4f176dcc2bcd286956c7fc4eea52586140cd3575/scripts/semantic_release_workflow.py#L84-L89

**Comment by PinrLi** [comment]
> hi @PinrLi I was testing on my end and faced an issue when we make any change to the root `pyproejct.toml` file, whenever any change is made to the root [`pyproejct.toml`](https://github.com/tattle-made/feluda/blob/main/pyproject.toml) file, the semantic version should be applied to `felulda`.
> 
> So in my test, I made [this commit](https://github.com/aatmanvaidya/feluda-versioning-check/commit/5f4d696e021c7b5236c9ac507c83819b7e84ede3) to the pyproject.toml file - then the [github action gave this output](https://github.com/aatmanvaidya/feluda-versioning-check/actions/runs/14436724272/job/40478989436). Here is the output of the script
> 
> ```shell
> No changes found for /home/runner/work/feluda-versioning-check/feluda-versioning-check/feluda. Skipping version bump.
> No changes found for /home/runner/work/feluda-versioning-check/feluda-versioning-check/operators/image_vec_rep_resnet. Skipping version bump.
> No changes found for /home/runner/work/feluda-versioning-check/feluda-versioning-check/operators/vid_vec_rep_clip. Skipping version bump.
> ```
> 
> I think this issue is coming form these lines in the code -
> 
> https://github.com/tattle-made/feluda/blob/4f176dcc2bcd286956c7fc4eea52586140cd3575/scripts/semantic_release_workflow.py#L84-L89

Hey I cloned your feluda-visioning-check, may I know how I should recreate this error? thanks

**Comment by aatmanvaidya** [comment]
> Hey I cloned your feluda-visioning-check, may I know how I should recreate this error? thanks

Hi, so I have a folder called [automated_testing](https://github.com/aatmanvaidya/feluda-versioning-check/tree/main/automate_testing), in this folder you will find a lot of `.sh` files that run end to end tests. But for this you will need [github cli](https://cli.github.com/) installed and then to replicate the error you will have to run - [02_feluda_pyproject_changes.sh](https://github.com/aatmanvaidya/feluda-versioning-check/blob/main/automate_testing/02_feluda_pyproject_changes.sh) file

if setting up github cli and doing things is an headache, do the following
1 Clone the repo
2. Make sure your latest [semantic version script](https://github.com/aatmanvaidya/feluda-versioning-check/blob/main/scripts/semantic_release_workflow.py) code is updated and pushed.
3. Then create a new branch, in this branch just make any change to the [root pyproject.toml](https://github.com/aatmanvaidya/feluda-versioning-check/blob/main/pyproject.toml) file.
4. Then create a PR from this branch, and merge the PR into main
5. After this in the Github Actions tab, you will see the action running and monitor the logs there, you will notice this error

**Comment by aatmanvaidya** [comment]
hi @PinrLi I understand that this small change and fix could be getting hetic, 

here is another suggestion - we don't make any changes to the script, instead we re-write the entire test. like in the test, we don't create a dummy git repo, we can discard it, we can just write 4-5 test functions that run unit test for some core semantic script function, this way we don't get into the mess of creating a dummy git repo etc.

**Comment by PinrLi** [comment]
> hi @PinrLi I understand that this small change and fix could be getting hetic,
> 
> here is another suggestion - we don't make any changes to the script, instead we re-write the entire test. like in the test, we don't create a dummy git repo, we can discard it, we can just write 4-5 test functions that run unit test for some core semantic script function, this way we don't get into the mess of creating a dummy git repo etc.

I was trying to modify slightly the semantic code, changing the package_root as a parameter and taht way I could overwrite the default path in the unittest, I think this method works without changing too much of the code, but I am still working on adjusting the unit test code.
`def _discover_packages(self, package_roots=["feluda"], operators_path="operators"):
  ...`

**Comment by aatmanvaidya** [comment]
> > hi @PinrLi I understand that this small change and fix could be getting hetic,
> > here is another suggestion - we don't make any changes to the script, instead we re-write the entire test. like in the test, we don't create a dummy git repo, we can discard it, we can just write 4-5 test functions that run unit test for some core semantic script function, this way we don't get into the mess of creating a dummy git repo etc.
> 
> I was trying to modify slightly the semantic code, changing the package_root as a parameter and taht way I could overwrite the default path in the unittest, I think this method works without changing too much of the code, but I am still working on adjusting the unit test code. `def _discover_packages(self, package_roots=["feluda"], operators_path="operators"): ...`

ahh noted, please carry on, I was just suggesting that as an alternative.

**Comment by PinrLi** [comment]
@aatmanvaidya Hey can you verify if the current version works? My unittest outputs:
`Ran 21 tests in 3.395s

OK (skipped=2)`

And I forked the versionning check repo and it looks good to me

**Comment by aatmanvaidya** [comment]
hi @PinrLi yes testing it on my end now

**Review by aatmanvaidya — APPROVED** [review]

**Comment by aatmanvaidya** [comment]
hi @PinrLi I have tested it out, both the unit test and automated tests work perfectly,

this is great work, thank you for your patience and consistent efforts on this, I will merge the PR by tomorrow.
